bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs

```diff
@@ -0,0 +1,283 @@
+		/// Message channel is invalid
+		InvalidChannel,
+		/// The max nonce for the type has been reached
+		MaxNonceReached,
+		/// Cannot convert location
+		InvalidAccountConversion,
+		/// Pallet is halted
+		Halted,
+		/// Message verification error,
+		Verification(VerificationError),
+		/// XCMP send failure
+		Send(SendError),
+		/// Message conversion error
+		ConvertMessage(ConvertMessageError),
+	}
+
+	#[derive(Clone, Encode, Decode, Eq, PartialEq, Debug, TypeInfo, PalletError)]
+	pub enum SendError {
+		NotApplicable,
+		NotRoutable,
+		Transport,
+		DestinationUnsupported,
+		ExceedsMaxMessageSize,
+		MissingArgument,
+		Fees,
+	}
+
+	impl<T: Config> From<XcmpSendError> for Error<T> {
+		fn from(e: XcmpSendError) -> Self {
+			match e {
+				XcmpSendError::NotApplicable => Error::<T>::Send(SendError::NotApplicable),
+				XcmpSendError::Unroutable => Error::<T>::Send(SendError::NotRoutable),
+				XcmpSendError::Transport(_) => Error::<T>::Send(SendError::Transport),
+				XcmpSendError::DestinationUnsupported =>
+					Error::<T>::Send(SendError::DestinationUnsupported),
+				XcmpSendError::ExceedsMaxMessageSize =>
+					Error::<T>::Send(SendError::ExceedsMaxMessageSize),
+				XcmpSendError::MissingArgument => Error::<T>::Send(SendError::MissingArgument),
+				XcmpSendError::Fees => Error::<T>::Send(SendError::Fees),
+			}
+		}
+	}
+
+	/// The nonce of the message been processed or not
+	#[pallet::storage]
+	pub type NonceBitmap<T: Config> = StorageMap<_, Twox64Concat, u128, u128, ValueQuery>;
+
+	/// The current operating mode of the pallet.
+	#[pallet::storage]
+	#[pallet::getter(fn operating_mode)]
+	pub type OperatingMode<T: Config> = StorageValue<_, BasicOperatingMode, ValueQuery>;
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Submit an inbound message originating from the Gateway contract on Ethereum
+		#[pallet::call_index(0)]
+		#[pallet::weight(T::WeightInfo::submit())]
+		pub fn submit(origin: OriginFor<T>, message: Message) -> DispatchResult {
+			let who = ensure_signed(origin.clone())?;
+			ensure!(!Self::operating_mode().is_halted(), Error::<T>::Halted);
+
+			// submit message to verifier for verification
+			T::Verifier::verify(&message.event_log, &message.proof)
+				.map_err(|e| Error::<T>::Verification(e))?;
+
+			// Decode event log into an Envelope
+			let envelope =
+				Envelope::try_from(&message.event_log).map_err(|_| Error::<T>::InvalidEnvelope)?;
+
+			// Verify that the message was submitted from the known Gateway contract
+			ensure!(T::GatewayAddress::get() == envelope.gateway, Error::<T>::InvalidGateway);
+
+			// Verify the message has not been processed
+			ensure!(!Nonce::<T>::get(envelope.nonce.into()), Error::<T>::InvalidNonce);
+
+			let origin_account_location = Self::account_to_location(who)?;
+
+			let (xcm, _relayer_reward) =
+				Self::do_convert(envelope.message, origin_account_location.clone())?;
+
+			// Todo: Deposit fee(in Ether) to RewardLeger which should cover all of:
```

Comment:
When will this be uncommented? Do we need to create a tracking issue?

Reply:
The reward functionality is added in https://github.com/paritytech/polkadot-sdk/pull/6578. Once that is merged, we can uncomment this. Added tracking issue: https://github.com/paritytech/polkadot-sdk/issues/7356

Reply:
In the interim, I've added a dummy trait until the new pallet is ready: https://github.com/paritytech/polkadot-sdk/pull/7402/commits/6f769ccdfd222654a26ce155f40468681d5ded28